File: backend/src/routes/uploads.js
import express from 'express';
import multer from 'multer';
import path from 'path';
import { fileURLToPath } from 'url';
import { verifyTokenMiddleware } from '../utils/auth.js';
import { getBucket, collections } from '../db.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const router = express.Router();

// Configure multer to buffer files in memory before piping to GridFS
const storage = multer.memoryStorage();

// File filter to allow common file types
const fileFilter = (req, file, cb) => {
  // Allow images, documents, and text files
  const allowedTypes = [
    'image/jpeg', 'image/png', 'image/gif', 'image/webp',
    'application/pdf',
    'application/msword', 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'application/vnd.ms-excel', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    'text/plain', 'text/csv',
    'application/json'
  ];
  
  if (allowedTypes.includes(file.mimetype)) {
    cb(null, true);
  } else {
    cb(new Error('File type not allowed'), false);
  }
};

// Configure multer with size limit (200MB)
const upload = multer({
  storage: storage,
  fileFilter: fileFilter,
  limits: {
    fileSize: 200 * 1024 * 1024 // 200MB in bytes
  }
});

// Upload endpoint - requires authentication
router.post('/upload', verifyTokenMiddleware, upload.array('files', 10), async (req, res) => {
  try {
    if (!req.files || req.files.length === 0) {
      return res.status(400).json({ error: 'No files uploaded' });
    }
    const bucket = getBucket();
    const metaCol = collections().consumers; // we'll store file refs per user in their document
    const email = req.user.email;

    const results = [];
    for (const file of req.files) {
      const filename = `${Date.now()}-${Math.round(Math.random()*1e9)}-${file.originalname}`;
      const uploadStream = bucket.openUploadStream(filename, { contentType: file.mimetype, metadata: { email } });
      await new Promise((resolve, reject) => {
        uploadStream.end(file.buffer, err => err ? reject(err) : resolve());
      });
      results.push({ filename, size: file.size, mimetype: file.mimetype });
    }

    res.json({ message: 'Files uploaded successfully', files: results, userEmail: email });
  } catch (error) {
    console.error('Upload error:', error);
    res.status(500).json({ error: 'Upload failed' });
  }
});

// Get user's uploaded files
router.get('/files', verifyTokenMiddleware, async (req, res) => {
  try {
    const email = req.user.email;
    const bucket = getBucket();
    const cursor = bucket.find({ 'metadata.email': email });
    const files = await cursor.toArray();
    const list = files.map(f => ({ filename: f.filename, size: f.length, uploadDate: f.uploadDate, id: f._id }));
    res.json({ files: list });
  } catch (error) {
    console.error('Error fetching files:', error);
    res.status(500).json({ error: 'Failed to fetch files' });
  }
});

// Download a file by filename
router.get('/files/:filename', verifyTokenMiddleware, async (req, res) => {
  try {
    const bucket = getBucket();
    const filename = req.params.filename;
    const file = await bucket.find({ filename, 'metadata.email': req.user.email }).next();
    if (!file) return res.status(404).json({ error: 'File not found' });
    res.setHeader('Content-Type', file.contentType || 'application/octet-stream');
    bucket.openDownloadStreamByName(filename).pipe(res);
  } catch (error) {
    console.error('Download error:', error);
    res.status(500).json({ error: 'Failed to download file' });
  }
});

// Error handling middleware for multer
router.use((error, req, res, next) => {
  if (error instanceof multer.MulterError) {
    if (error.code === 'LIMIT_FILE_SIZE') {
      return res.status(400).json({ error: 'File too large. Maximum size is 200MB.' });
    }
    if (error.code === 'LIMIT_FILE_COUNT') {
      return res.status(400).json({ error: 'Too many files. Maximum is 10 files.' });
    }
  }
  
  if (error.message === 'File type not allowed') {
    return res.status(400).json({ error: 'File type not allowed' });
  }
  
  res.status(500).json({ error: 'Upload failed' });
});

export default router;

File: backend_python/src/routes/uploads.py
import os
import sys
import logging
from datetime import datetime
from typing import List
from fastapi import APIRouter, HTTPException, Depends, status, UploadFile, File
from fastapi.responses import StreamingResponse
from io import BytesIO

from ..utils.auth import verify_token_middleware

try:  # Prefer eager import, but keep details if it fails
    from minio import Minio  # type: ignore
    _MINIO_IMPORT_ERROR: str | None = None
except Exception as _e:  # pragma: no cover - tests patch client, import may be absent
    Minio = None  # type: ignore
    _MINIO_IMPORT_ERROR = repr(_e)

logger = logging.getLogger(__name__)

router = APIRouter()

# Allowed file types
ALLOWED_TYPES = {
    'image/jpeg', 'image/png', 'image/gif', 'image/webp',
    'application/pdf',
    'application/msword', 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'application/vnd.ms-excel', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    'text/plain', 'text/csv',
    'application/json'
}

# File size limit (200MB)
MAX_FILE_SIZE = 200 * 1024 * 1024


def validate_file(file: UploadFile) -> None:
    """Validate uploaded file"""
    if file.content_type not in ALLOWED_TYPES:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="File type not allowed"
        )
    
    # Note: FastAPI doesn't provide file size during upload
    # Size validation would need to be done during reading


# --- MinIO helpers -----------------------------------------------------------
def get_minio_client():
        """Create a MinIO client from environment variables.

        Defaults:
            - endpoint: 127.0.0.1:8803
            - access/secret: minioadmin/minioadmin
            - secure: False
        """
        # Lazy import in case the module was installed after process start
        global Minio, _MINIO_IMPORT_ERROR
        if Minio is None:
            try:
                from minio import Minio as _Minio  # type: ignore
                Minio = _Minio  # type: ignore
                _MINIO_IMPORT_ERROR = None
            except Exception as e:  # pragma: no cover
                _MINIO_IMPORT_ERROR = repr(e)
                # Log helpful diagnostics
                logger.error("MinIO import failed: %s | python=%s", _MINIO_IMPORT_ERROR, sys.executable)
                raise HTTPException(
                    status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                    detail="MinIO client unavailable: install the 'minio' package on the backend or configure storage."
                )
        endpoint = os.getenv("MINIO_ENDPOINT", "127.0.0.1:8803")
        access_key = os.getenv("MINIO_ACCESS_KEY", "minio")
        secret_key = os.getenv("MINIO_SECRET_KEY", "minio8888")
        secure = os.getenv("MINIO_SECURE", "false").lower() == "true"
        # The tests will monkeypatch this function to return a fake client
        return Minio(endpoint, access_key=access_key, secret_key=secret_key, secure=secure)  # type: ignore


def get_bucket_name() -> str:
    return os.getenv("MINIO_BUCKET", "hcp")


@router.post("/upload")
async def upload_files(
    files: List[UploadFile] = File(...),
    user: dict = Depends(verify_token_middleware)
):
    """Upload files to MinIO S3 bucket with user email prefix."""
    if not files:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="No files uploaded"
        )
    
    if len(files) > 10:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Too many files. Maximum is 10 files."
        )
    
    try:
        client = get_minio_client()
        bucket_name = get_bucket_name()
        user_email = user.get("email") or "unknown"

        # Ensure bucket exists
        try:
            if hasattr(client, "bucket_exists") and not client.bucket_exists(bucket_name):  # type: ignore
                client.make_bucket(bucket_name)  # type: ignore
        except Exception:
            # If check fails, continue and let put_object surface errors
            pass

        results = []

        for file in files:
            validate_file(file)

            content = await file.read()
            if len(content) > MAX_FILE_SIZE:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"File {file.filename} too large. Maximum size is 200MB."
                )

            timestamp = int(datetime.now().timestamp() * 1000)
            random_part = os.urandom(4).hex()
            filename = f"{timestamp}-{random_part}-{file.filename}"
            object_name = f"{user_email}/{filename}"

            # Upload to MinIO
            data_stream = BytesIO(content)
            size = len(content)
            content_type = file.content_type or "application/octet-stream"
            client.put_object(bucket_name, object_name, data_stream, size, content_type=content_type)  # type: ignore

            results.append({
                "filename": filename,
                "originalName": file.filename,
                "size": size,
                "mimetype": content_type,
                "key": object_name,
            })

        return {
            "message": "Files uploaded successfully",
            "files": results,
            "userEmail": user_email,
            "bucket": bucket_name,
        }

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Upload error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Upload failed"
        )


@router.get("/files")
async def get_user_files(user: dict = Depends(verify_token_middleware)):
    """Get user's uploaded files"""
    try:
        client = get_minio_client()
        bucket_name = get_bucket_name()
        user_email = user.get("email") or "unknown"

        # List objects with user's prefix
        prefix = f"{user_email}/"
        objs = client.list_objects(bucket_name, prefix=prefix, recursive=True)  # type: ignore

        file_list = []
        for obj in objs:
            # obj.object_name like 'email/filename'
            name = obj.object_name.split("/", 1)[1] if "/" in obj.object_name else obj.object_name
            file_list.append({
                "filename": name,
                "size": getattr(obj, "size", None),
                "uploadDate": getattr(obj, "last_modified", None),
                "key": obj.object_name,
            })

        return {"files": file_list}

    except Exception as e:
        logger.error(f"Error fetching files: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to fetch files"
        )


@router.get("/files/{filename}")
async def download_file(
    filename: str,
    user: dict = Depends(verify_token_middleware)
):
    """Download a file by filename"""
    try:
        client = get_minio_client()
        bucket_name = get_bucket_name()
        user_email = user.get("email") or "unknown"

        object_name = f"{user_email}/{filename}"
        try:
            obj = client.get_object(bucket_name, object_name)  # type: ignore
        except Exception:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="File not found")

        # Read entire content
        try:
            content = obj.read()  # type: ignore
        except Exception:
            # Some clients return a stream-like object with read()
            content = obj.read() if hasattr(obj, "read") else bytes()

        def iterfile():
            yield content

        return StreamingResponse(
            iterfile(),
            media_type="application/octet-stream",
            headers={"Content-Disposition": f"attachment; filename={filename}"}
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Download error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to download file"
        )


@router.get("/diag")
async def uploads_diagnostics():
    """Lightweight diagnostics to debug upload 500s without changing code."""
    return {
        "minioImport": "ok" if Minio is not None else f"missing: {_MINIO_IMPORT_ERROR}",
        "python": sys.executable,
        "env": {
            "MINIO_ENDPOINT": os.getenv("MINIO_ENDPOINT", "(unset)"),
            "MINIO_SECURE": os.getenv("MINIO_SECURE", "(unset)"),
            "MINIO_BUCKET": os.getenv("MINIO_BUCKET", "(unset)"),
        },
    }
